File: Algoritmos de ordenamiento/py/12 BinaryTree.py
class Node:
    def __init__(self, value):
        self.data = value
        self.left = None
        self.right = None

def create_node(value):
    new_node = Node(value)
    return new_node

def insert_node(root, value):
    if root is None:
        return create_node(value)

    if value < root.data:
        root.left = insert_node(root.left, value)
    else:
        root.right = insert_node(root.right, value)

    return root

def in_order_traversal(root, array, index):
    if root is None:
        return

    in_order_traversal(root.left, array, index)
    array[index[0]] = root.data
    index[0] += 1
    in_order_traversal(root.right, array, index)

def binary_tree_sort(array):
    size = len(array)
    root = None

    for i in range(size):
        root = insert_node(root, array[i])

    index = [0]
    in_order_traversal(root, array, index)

array = [2, 3, 1, 923, -23, 9128, 0, 10, 20, 12]

print("\n== Estos son tus valores == \n")
print(*array, sep="\t")

print("\n\n== Ordenando... ==\n")

binary_tree_sort(array)

for num in array:
    print(num, end="\t")
print()

print("\n== Estos son tus valores ordenados ==\n")
print(*array, sep="\t")

print("\n")

input("Presiona Enter para continuar...")

File: Algoritmos de ordenamiento/cpp/12 BinaryTree.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
};

Node* createNode(int value) {
    Node* newNode = new Node();
    if (newNode != nullptr) {
        newNode->data = value;
        newNode->left = nullptr;
        newNode->right = nullptr;
    }
    return newNode;
}

Node* insertNode(Node* root, int value) {
    if (root == nullptr) {
        return createNode(value);
    }

    if (value < root->data) {
        root->left = insertNode(root->left, value);
    } else {
        root->right = insertNode(root->right, value);
    }

    return root;
}

void inOrderTraversal(Node* root, int* array, int* index) {
    if (root == nullptr) {
        return;
    }

    inOrderTraversal(root->left, array, index);
    array[(*index)++] = root->data;
    inOrderTraversal(root->right, array, index);
}

void binaryTreeSort(int array[], size_t size) {
    Node* root = nullptr;

    for (size_t i = 0; i < size; i++) {
        root = insertNode(root, array[i]);
    }

    int index = 0;
    inOrderTraversal(root, array, &index);

    delete root;
}

int main() {
    system("title Algoritmos de ordenamientos de datos en C++ / Binary Tree Sort && cls");

    int array[10] = {2, 3, 1, 923, -23, 9128, 0, 10, 20, 12};
    size_t size = sizeof(array) / sizeof(array[0]);

    cout << "\n== Estos son tus valores == \n\n";
    for (int i = 0; i < size; i++) {
        cout << array[i] << "    ";
    }
    cout << endl;

    cout << "\n\n== Ordenando... ==\n\n";

    binaryTreeSort(array, size);

    cout << "\n== Estos son tus valores ordenados ==\n\n";
    for (int i = 0; i < size; i++) {
        cout << array[i] << "    ";
    }

    cout << endl << endl;

    system("pause");
    system("cls");
}
